crates/ide-diagnostics/src/handlers/no_such_field.rs

```diff
@@ -34,11 +35,78 @@ pub(crate) fn no_such_field(ctx: &DiagnosticsContext<'_>, d: &hir::NoSuchField)
             node,
         )
         .stable()
-        .with_fixes(fixes(ctx, d))
+        .with_fixes(no_such_field_fixes(ctx, d))
     }
 }
 
-fn fixes(ctx: &DiagnosticsContext<'_>, d: &hir::NoSuchField) -> Option<Vec<Assist>> {
+fn field_is_private_fixes(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::NoSuchField,
+) -> Option<Vec<Assist>> {
+    let root = ctx.sema.db.parse_or_expand(d.field.file_id);
+    match &d.field.value.to_node(&root) {
+        Either::Left(node) => private_record_expr_field_fixes(&ctx.sema, node),
+        _ => None,
+    }
+}
+fn private_record_expr_field_fixes(
+    sema: &Semantics<'_, RootDatabase>,
+    record_expr_field: &ast::RecordExprField,
+) -> Option<Vec<Assist>> {
+    let record_lit = ast::RecordExpr::cast(record_expr_field.syntax().parent()?.parent()?)?;
+    let def_id = sema.resolve_variant(record_lit)?;
+    let def_file_id;
+    let record_fields = match def_id {
+        hir::VariantDef::Struct(s) => {
+            let source = s.source(sema.db)?;
+            def_file_id = source.file_id;
+            let fields = source.value.field_list()?;
+            record_field_list(fields)?
+        }
+        hir::VariantDef::Union(u) => {
+            let source = u.source(sema.db)?;
+            def_file_id = source.file_id;
+            source.value.record_field_list()?
+        }
+        hir::VariantDef::Variant(e) => {
+            let source = e.source(sema.db)?;
+            def_file_id = source.file_id;
+            let fields = source.value.field_list()?;
+            record_field_list(fields)?
+        }
+    };
+    let def_file_id = def_file_id.original_file(sema.db);
+
+    let field_definition = find_field(&record_fields, record_expr_field)?;
+
+    let source_change = SourceChange::from_text_edit(
+        def_file_id.file_id(sema.db),
+        TextEdit::insert(field_definition.syntax().text_range().start(), "pub ".into()),
```

Comment:
The text range here might belong to a macro file here, so we need to upmap that out of macros (and use the upmapped file id from there). Otherwise things will break